### ISSUE TITLE
feature: add fluent API for templates

### ISSUE THREAD

**Issue by pvojtechovsky** [INRIA/spoon#1458]
```java
class _Builder_ extends AbstractBuilder implements ThisIFace, ThatIFace {
  private  _Proxy_ delegate;
  public _Builder_(_Proxy_ proxy) {
    this.proxy = proxy;
  }
  @Local
  public _Builder_ _propertyName_(_PropertyType_ _propertyName_) {
     delegate._setterName_(_propertyName_);
     return this;
  }
  @Local
  public static CtClass<?> createFluentBuilder(String builderQName, CtTypeReference<?> proxyType) {
      return new SubstitutionRequest(proxyType.getFactory(), _Builder_ .class)
          .excludeAllMembersWithAnnotation(Local.class)
          .substituteClass(_Proxy_ .class, proxyType)
          .createType(builderQName)
  }
  @Local
  public static List<CtTypeMember> addFluentSetter(CtClass<?> builder, CtMethod<?> setterMethod) {
      return new SubstitutionRequest(setterMethod.getFactory(), _Builder_ .class)
          .excludeAll().includeMethod("_propertyName_")
          .substituteClass(_Proxy_ .class, setterMethod.getDeclaringType())
          .substituteName("_setterName_", setterMethod.getSimpleName())
          .substituteName("_propertyName", Utils.lowerCaseFirst(setterMethod.getSimpleName().substring(3)))
          .applyToType(builder)
  }
}
```
and usage
```java
CtClass<?> newBuilderClass = _Builder_ .createFluentBuilder("a.b.c.FluentSetterForX", typeRefToTypeX);
for(CtMethod m : Utils.getAllSetters(typeRefToTypeX)) {
 _Builder_.addFluentSetter(newBuilderClass, m);
}
```
WDYT?

**Comment by surli** [comment]
I like the idea of a more explicit API for substitution inside the template. However in this example, why the `Local` annotation are not excluded by default? And why do you use one on `_propertyName_`? It is a mistake or it is on purpose?

**Comment by pvojtechovsky** [comment]
> why the Local annotation are not excluded by default? And why do you use one on _propertyName_? It is a mistake or it is on purpose?

It is on purpose. Note that type `_Builder_` defines 2 templates! (Not one like it was always till now!). First template constructed by method `createFluentBuilder` is used to generate target class itself. 
The second template constructed by method `addFluentSetter` is used to generate one method into existing (before generated) type.

Therefore the first template must NOT include method `_propertyName_`, while second template includes ONLY method `_propertyName_`.

Why to put more template definitions (`createFluentBuilder`  and `addFluentSetter`) into one type '_Builder_'? Why not to create two template types 'TypeBuilder' and `MethodBuilder`?
Because there can be many other type members (fields, methods, subclasses), which should be generated by `createFluentBuilder` and which are needed to compile the template method. The putting all together avoids copying of template code and easier refactoring and understanding of complex templates.

Note: that the TemplateBuilder does not understand meaning of annotation `@Local`. It is same annotation like any other from TemplateBuilder point of view. From this point of view is the usage of `@Local` annotation in example above little bit confusing. There might be any client specific annotation, which fits to purpose to organize parts of her/his templates. I personally use my own `@Tag` annotation.

**Comment by pvojtechovsky** [comment]
Will be solved by #1686

**Comment by pvojtechovsky** [comment]
The ideas of this issue are obsolete. PR #1686 has already better implementation. Closing this one
